File: src/MPK_SOEM/src/cia402.cpp
#include "mpk_soem/cia402.hpp"

#include <algorithm>
#include <cctype>
#include <stdexcept>

namespace mpk_soem::cia402
{

DriveState decode_status_word(uint16_t status_word)
{
  const uint16_t masked_4f = status_word & 0x004F;
  const uint16_t masked_6f = status_word & 0x006F;

  if (masked_4f == 0x0000) {
    return DriveState::NotReadyToSwitchOn;
  }
  if (masked_4f == 0x0040) {
    return DriveState::SwitchOnDisabled;
  }
  if (masked_6f == 0x0021) {
    return DriveState::ReadyToSwitchOn;
  }
  if (masked_6f == 0x0023) {
    return DriveState::SwitchedOn;
  }
  if (masked_6f == 0x0027) {
    return DriveState::OperationEnabled;
  }
  if (masked_6f == 0x0007) {
    return DriveState::QuickStopActive;
  }
  if (masked_4f == 0x000F) {
    return DriveState::FaultReactionActive;
  }
  if (masked_4f == 0x0008) {
    return DriveState::Fault;
  }
  return DriveState::Unknown;
}

const char * to_string(DriveState state)
{
  switch (state) {
    case DriveState::NotReadyToSwitchOn:
      return "Not ready to switch on";
    case DriveState::SwitchOnDisabled:
      return "Switch on disabled";
    case DriveState::ReadyToSwitchOn:
      return "Ready to switch on";
    case DriveState::SwitchedOn:
      return "Switched on";
    case DriveState::OperationEnabled:
      return "Operation enabled";
    case DriveState::QuickStopActive:
      return "Quick stop active";
    case DriveState::FaultReactionActive:
      return "Fault reaction active";
    case DriveState::Fault:
      return "Fault";
    case DriveState::Unknown:
    default:
      return "Unknown";
  }
}

const char * to_string(Mode mode)
{
  switch (mode) {
    case Mode::NoMode:
      return "No mode";
    case Mode::ProfilePosition:
      return "Profile position";
    case Mode::ProfileVelocity:
      return "Profile velocity";
    case Mode::Homing:
      return "Homing";
    case Mode::InterpolatedPosition:
      return "Interpolated position";
    case Mode::Csp:
      return "CSP";
    case Mode::Csv:
      return "CSV";
    case Mode::Cst:
      return "CST";
    default:
      return "Unknown";
  }
}

bool is_operation_enabled(uint16_t status_word)
{
  return decode_status_word(status_word) == DriveState::OperationEnabled;
}

bool is_fault(uint16_t status_word)
{
  const auto state = decode_status_word(status_word);
  return state == DriveState::Fault || state == DriveState::FaultReactionActive;
}

uint16_t next_enable_control_word(uint16_t status_word, uint16_t previous_control_word)
{
  switch (decode_status_word(status_word)) {
    case DriveState::Fault:
      return previous_control_word == control_word::fault_reset ?
        control_word::disable_voltage : control_word::fault_reset;
    case DriveState::FaultReactionActive:
      return control_word::disable_voltage;
    case DriveState::SwitchOnDisabled:
      return control_word::shutdown;
    case DriveState::ReadyToSwitchOn:
      return previous_control_word == control_word::switch_on ?
        control_word::enable_operation : control_word::switch_on;
    case DriveState::SwitchedOn:
    case DriveState::OperationEnabled:
      return control_word::enable_operation;
    case DriveState::QuickStopActive:
      return control_word::shutdown;
    case DriveState::NotReadyToSwitchOn:
    case DriveState::Unknown:
    default:
      return control_word::disable_voltage;
  }
}

Mode parse_mode(const std::string & mode_name)
{
  std::string m;
  m.reserve(mode_name.size());
  std::transform(mode_name.begin(), mode_name.end(), std::back_inserter(m), [](unsigned char c) {
    return static_cast<char>(std::tolower(c));
  });

  if (m == "csp" || m == "8") {
    return Mode::Csp;
  }
  if (m == "csv" || m == "9") {
    return Mode::Csv;
  }
  if (m == "pp" || m == "profile-position" || m == "profile_position" || m == "1") {
    return Mode::ProfilePosition;
  }
  if (m == "pv" || m == "profile-velocity" || m == "profile_velocity" || m == "3") {
    return Mode::ProfileVelocity;
  }
  if (m == "none" || m == "idle" || m == "0") {
    return Mode::NoMode;
  }
  throw std::invalid_argument("unsupported CiA402 mode: " + mode_name);
}

}  // namespace mpk_soem::cia402

File: src/multi_motor/multi_motor_control/cia402.py
# -*- coding: utf-8 -*-
"""
cia402.py
---------
CiA 402 state machine / mode constants + Status-Word parser.
Self-contained: no ROS, no Qt imports - safe to unit-test.
"""

from dataclasses import dataclass


# ---------- Control Word (object 0x6040) ----------
class CW:
    DISABLE_VOLTAGE  = 0x0000
    QUICK_STOP       = 0x0002
    SHUTDOWN         = 0x0006
    SWITCH_ON        = 0x0007
    ENABLE_OPERATION = 0x000F
    FAULT_RESET      = 0x0080   # rising edge on bit 7
    HALT_BIT         = 0x0100
    NEW_SET_POINT    = 0x0010   # PP mode: bit 4 rising edge
    CHANGE_IMMEDIATE = 0x0020   # PP mode: bit 5


# ---------- Modes of Operation (object 0x6060/0x6061) ----------
class MODE:
    NO_MODE           = 0
    PROFILE_POSITION  = 1
    PROFILE_VELOCITY  = 3
    PROFILE_TORQUE    = 4
    HOMING            = 6
    INTERPOLATED_POS  = 7
    CSP               = 8
    CSV               = 9
    CST               = 10

    NAMES = {
        0:  "No Mode",
        1:  "Profile Position",
        3:  "Profile Velocity",
        4:  "Profile Torque",
        6:  "Homing",
        7:  "Interpolated Pos",
        8:  "CSP",
        9:  "CSV",
        10: "CST",
    }


# ---------- Status Word (object 0x6041) ----------
_STATE_MASK     = 0x006F  # bits 0,1,2,3,5,6
# Canonical masked values for each CiA 402 state:
_STATE_NAMES = [
    ("Not Ready To Switch On", 0x0000, 0x004F),
    ("Switch On Disabled",     0x0040, 0x004F),
    ("Ready To Switch On",     0x0021, 0x006F),
    ("Switched On",            0x0023, 0x006F),
    ("Operation Enabled",      0x0027, 0x006F),
    ("Quick Stop Active",      0x0007, 0x006F),
    ("Fault Reaction Active",  0x000F, 0x004F),
    ("Fault",                  0x0008, 0x004F),
]


@dataclass
class DriveStatus:
    raw: int
    state: str
    ready_to_switch_on: bool
    switched_on: bool
    operation_enabled: bool
    fault: bool
    voltage_enabled: bool
    quick_stop: bool          # bit 5 = 1 means NOT in quick-stop
    switch_on_disabled: bool
    warning: bool
    target_reached: bool
    internal_limit: bool


def parse_status_word(sw: int) -> DriveStatus:
    """Decode a 16-bit Status Word into named flags + CiA 402 state."""
    sw &= 0xFFFF
    state = "Unknown"
    for name, value, mask in _STATE_NAMES:
        if (sw & mask) == value:
            state = name
            break
    return DriveStatus(
        raw=sw,
        state=state,
        ready_to_switch_on = bool(sw & (1 << 0)),
        switched_on        = bool(sw & (1 << 1)),
        operation_enabled  = bool(sw & (1 << 2)),
        fault              = bool(sw & (1 << 3)),
        voltage_enabled    = bool(sw & (1 << 4)),
        quick_stop         = bool(sw & (1 << 5)),
        switch_on_disabled = bool(sw & (1 << 6)),
        warning            = bool(sw & (1 << 7)),
        target_reached     = bool(sw & (1 << 10)),
        internal_limit     = bool(sw & (1 << 11)),
    )


def next_control_word(current_sw: int) -> int:
    """
    Given the current Status Word, return the Control Word that
    advances the CiA 402 state machine one step toward
    Operation Enabled. Returns CW.ENABLE_OPERATION once the
    drive is already in Operation Enabled.
    """
    s = parse_status_word(current_sw)
    if s.fault:
        return CW.FAULT_RESET
    if s.switch_on_disabled:
        return CW.SHUTDOWN            # -> Ready to Switch On
    if s.ready_to_switch_on and not s.switched_on:
        return CW.SWITCH_ON           # -> Switched On
    if s.switched_on and not s.operation_enabled:
        return CW.ENABLE_OPERATION    # -> Operation Enabled
    return CW.ENABLE_OPERATION
